tensorflow_probability/python/sts/trigonometric.py

```diff
@@ -0,0 +1,225 @@
+          value=observation_noise_scale,
+          name='observation_noise_scale',
+          dtype=dtype)
+
+      num_frequencies = len(selected_frequencies)
+
+      observation_matrix = tf.tile(
+          input=tf.constant([[1., 0.]], dtype=dtype),
+          multiples=[1, num_frequencies])
+
+      transition_matrix = build_trigonometric_transition_matrix(
+          period=period, selected_frequencies=selected_frequencies, dtype=dtype)
+
+      self._drift_scale = drift_scale
+      self._observation_noise_scale = observation_noise_scale
+      self._period = period
+      self._selected_frequencies = selected_frequencies
+
+      super(TrigonometricStateSpaceModel, self).__init__(
+          num_timesteps=num_timesteps,
+          transition_matrix=transition_matrix,
+          transition_noise=tfd.MultivariateNormalDiag(
+              scale_diag=(drift_scale[..., tf.newaxis] *
+                          tf.ones([2 * num_frequencies], dtype=dtype)),
+              name='transition_noise'),
+          observation_matrix=observation_matrix,
+          observation_noise=tfd.MultivariateNormalDiag(
+              scale_diag=observation_noise_scale[..., tf.newaxis],
+              name='observation_noise'),
+          initial_state_prior=initial_state_prior,
+          initial_step=initial_step,
+          allow_nan_stats=allow_nan_stats,
+          validate_args=validate_args,
+          name=name)
+
+  @property
+  def drift_scale(self):
+    """Standard deviation of the drift in the cyclic effects."""
+    return self._drift_scale
+
+  @property
+  def observation_noise_scale(self):
+    """Standard deviation of the observation noise."""
+    return self._observation_noise_scale
+
+  @property
+  def period(self):
+    """The seasonal period."""
+    return self._period
+
+  @property
+  def selected_frequencies(self):
+    """Chosen frequency indices to include."""
+    return self._selected_frequencies
+
+
+def build_trigonometric_transition_matrix(period, selected_frequencies, dtype):
+  """Build the transition matrix for a TrigonometricStateSpaceModel"""
+
+  two_pi = tf.constant(2. * np.pi, dtype=dtype)
+  frequencies = two_pi * selected_frequencies / period
+  num_frequencies = len(selected_frequencies)
+
+  sin_frequencies = tf.sin(frequencies)
+  cos_frequencies = tf.cos(frequencies)
+
+  trigonometric_values = tf.stack(
+      [cos_frequencies, sin_frequencies, -sin_frequencies, cos_frequencies],
+      axis=-1)
+
+  transition_matrix = tf.linalg.LinearOperatorBlockDiag(
+      [
+          tf.linalg.LinearOperatorFullMatrix(
+              matrix=tf.reshape(trigonometric_values[i], [2, 2]),
+              is_square=True
+          )
+          for i in range(num_frequencies)
+      ]
+  )
+
+  return transition_matrix
+
+
+class Trigonometric(StructuralTimeSeries):
+  """Formal representation of a trigonometric seasonal effects model."""
+
+  def __init__(self,
+               period,
+               selected_frequencies,
```

Comment:
I've been trying to follow [Andrew Harvey's book](https://www.cambridge.org/core/books/forecasting-structural-time-series-models-and-the-kalman-filter/CE5E112570A56960601760E786A5E631) where he writes e.g.:

> ... a set of trigonometric terms at the seasonal frequencies, lambda[j] = 2 * pi * j / s,  j = 1, ..., [s/2] ...

so I had denoted `lambda` with `frequency` and `j` with `selected_frequencies`. 

Probably just me, but I wasn't sure with BSTS what the `frequencies` argument was, i.e. is it `j` or `lambda` using Harvey's notation. I guessed it must be `j` because otherwise you don't need the `period` argument, I think.

Anyway, I agree `frequency_multipliers` is a good name - again it describes exactly what this does - so I've changed this in 3d867e80ef0232a480db351b7291852aac085150

Reply:
Sounds good, thanks!